### ISSUE TITLE
List of dynamic Enums can contain anything

### ISSUE THREAD

**Issue by k2bd** [enthought/traits#504]
Traits 5.1.2-1

MWE:

```python
from traits.api import HasTraits, Enum, List


class MyClass(HasTraits):
    vals = ["a", "b"]

    list_enum = List(Enum(values="vals"))


my_obj = MyClass()

my_obj.list_enum = ["c", None, MyClass()]
my_obj.list_enum.append(1)

print(my_obj.list_enum)
# ['c', None, <__main__.MyClass object at 0x1157010a0>, 1]
```

**Comment by rkern** [comment]
Yes, the dynamic-valued `Enum` has an odd implementation. It basically disables the `validate` method and handles the validation in an ad hoc manner in the setter, which is not what is used when it's an inner trait inside of `List`.

**Comment by pzahemszky** [comment]
As Robert was saying, this issue only affects dynamic-valued Enums:

```python
from traits.api import HasTraits, Enum, List

vals = ["a", "b"]


class MyClass(HasTraits):

    list_enum = List(Enum(vals))


my_obj = MyClass()

my_obj.list_enum = ["c", None, MyClass()]
# traits.trait_errors.TraitError: Each element of the 'list_enum' trait
# of a MyClass instance must be 'a' or 'b', but a value of 'c'
# <class 'str'> was specified
```

I slightly changed the title to show this.

**Comment by mdickinson** [comment]
This was fixed in #1388.
